### PR TITLE
Added conformance tests for OffscreenCanvas

### PR DESCRIPTION
**Pull request by xidachen**
The spec for OffscreenCanvas is here:
https://html.spec.whatwg.org/multipage/scripting.html#the-offscreencanvas-interface

In this pull request, we add two simple tests only, which is to ensure that webgl and webgl2 context creation is correct.

Let's put this pull request on hold for few days until this change lands because currently creating webgl2 context is not supported in chromium yet and the test will fail:
https://codereview.chromium.org/2509193002/

@kenrussell @zhenyao Please review.

### CODE REVIEW

**Comment by kenrussell** [comment]
Please hold off merging this for the moment. We're trying to take the WebGL 2.0.0 snapshot.

**Comment by xidachen** [comment]
@kenrussell @zhenyao PTAL. All these tests pass locally on a linux workstation once this CL lands:
https://codereview.chromium.org/2531853003/

In the subsequent commits, I have added several tests, so right now there are tests for:
1. getContext
2. Validating all WebGL/WebGL2 APIs
3. Context lost tests
4. Context restore tests.
5. All the above tests in a worker.

@junov FYI.

**Comment by xidachen** [comment]
Gentle ping.
@kenrussell @zhenyao Could you take a look at these tests?

File: sdk/tests/conformance/00_test_list.txt [hunk 1]
```diff
@@ -7,6 +7,7 @@ glsl/00_test_list.txt
 limits/00_test_list.txt
 misc/00_test_list.txt
 --min-version 1.0.2 ogles/00_test_list.txt
+offscreencanvas/00_test_list.txt
```
Comment (kenrussell) [inline sdk/tests/conformance/00_test_list.txt]: Please add the prefix "--min-version 1.0.4". This is mainly for bookkeeping, but it's still useful.

File: sdk/tests/conformance/offscreencanvas/context-lost-worker.html [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+        self.postMessage("Test failed");
+
+    imageData = new ImageData(1, 1);
+    float32array = new Float32Array(1);
+    int32array = new Int32Array(1);
+
+    // Functions returning void should return immediately.
+    // This is untestable, but we can at least be sure they cause no errors
+    // and the codepaths are exercised.
+    if (!compareGLError(gl.NO_ERROR, "gl.activeTexture(gl.TEXTURE0)") ||
```
Comment (kenrussell) [inline sdk/tests/conformance/offscreencanvas/context-lost-worker.html]: This is an awful lot of duplicated code compared to the similar test for canvases on the main thread. Can you please figure out how to share the code? In the past, to share code between WebGL 1.0 and 2.0 tests, we refactored the JavaScript portions of the tests into sdk/tests/js/tests/ and had the HTML files refer to that code. Can the similar pattern be done to share the code for regular canvas tests and OffscreenCanvas?

File: sdk/tests/conformance/offscreencanvas/context-lost.html [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+  });
+}
+
+function testFunctionsThatReturnNULL(tests) {
+  tests.forEach(function(test) {
+      shouldBeNull(test);
+  });
+}
+
+function testLostContext()
```
Comment (kenrussell) [inline sdk/tests/conformance/offscreencanvas/context-lost.html]: Again, this is a lot of duplicated code. Please figure out some way to share it. Thanks.

**Comment by kenrussell** [comment]
Also, please confirm that these tests pass in some other browser which doesn't yet support OffscreenCanvas. Thanks.

**Comment by xidachen** [comment]
@kenrussell @zhenyao In the last commit, I have refactored all the tests to avoid duplicate code. I know we should hold the merge after taking WebGL 2.0 snapshot, but could you take a look? Thank you.

File: sdk/tests/conformance/offscreencanvas/context-creation.html [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+<body>
+<div id="description"></div>
+<div id="console"></div>
+<script>
+"use strict";
+description("This test ensures that the WebGL context can be created on an OffscreenCanvas.");
+
+var successfullyParsed = false;
+if (!window.OffscreenCanvas) {
+  successfullyParsed = true;
```
Comment (zhenyao) [inline sdk/tests/conformance/offscreencanvas/context-creation.html]: You really don't need set successfullyParsed to true as finishTest() does that.

Also, it means you parse the html to the end, so even if the test fails, this should still be set to true. (which finishTest() did)

File: sdk/tests/conformance/offscreencanvas/context-creation.html [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  finishTest();
+}
+
+if (contextCreation('webgl')) {
+  successfullyParsed = true;
```
Comment (zhenyao) [inline sdk/tests/conformance/offscreencanvas/context-creation.html]: Same here

File: sdk/tests/conformance/offscreencanvas/context-lost-restored-worker.html [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+<script src="../../js/js-test-pre.js"></script>
+<script src="../../js/webgl-test-utils.js"></script>
+<script>
+function init()
+{
+  description("Tests behavior under a restored context for OffscreenCanvas in a worker.");
+
+  var successfullyParsed = false;
+  if (!window.OffscreenCanvas) {
+    successfullyParsed = true;
```
Comment (zhenyao) [inline sdk/tests/conformance/offscreencanvas/context-lost-restored-worker.html]: Same here.

File: sdk/tests/conformance/offscreencanvas/context-lost-restored-worker.html [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    finishTest();
+  }
+
+  var worker = new Worker('context-lost-restored-worker.js');
+  worker.postMessage("Start worker");
+  worker.onmessage = function(e) {
+    if (e.data == "Test passed") {
+      testPassed("All tests have passed");
+      successfullyParsed = true;
```
Comment (zhenyao) [inline sdk/tests/conformance/offscreencanvas/context-lost-restored-worker.html]: Same here

File: sdk/tests/conformance/offscreencanvas/context-lost-restored.html [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+<script src="../../js/tests/canvas-tests-utils.js"></script>
+<script>
+
+function init()
+{
+    description("Tests behavior under a restored context for OffscreenCanvas.");
+
+    var successfullyParsed = false;
+    if (!window.OffscreenCanvas) {
+      successfullyParsed = true;
```
Comment (zhenyao) [inline sdk/tests/conformance/offscreencanvas/context-lost-restored.html]: Same here

File: sdk/tests/conformance/offscreencanvas/context-lost-restored.html [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+        // restore the context after this event has exited.
+        setTimeout(function() {
+            // we didn't call prevent default so we should not be able to restore the context
+            if (!compareGLError(gl.INVALID_OPERATION, "WEBGL_lose_context.restoreContext()")) {
+                testFailed("Some test failed");
+                finishTest();
+            }
+            testLosingAndRestoringContext().then(function() {
+                testPassed("Test passed");
+                successfullyParsed = true;
```
Comment (zhenyao) [inline sdk/tests/conformance/offscreencanvas/context-lost-restored.html]: Same here.

File: sdk/tests/conformance/offscreencanvas/context-lost-worker.html [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+<script src="../../js/js-test-pre.js"></script>
+<script src="../../js/webgl-test-utils.js"></script>
+<script>
+function init()
+{
+  description("Tests behavior under a lost context for OffscreenCanvas in a worker");
+
+  var successfullyParsed = false;
+  if (!window.OffscreenCanvas) {
+    successfullyParsed = true;
```
Comment (zhenyao) [inline sdk/tests/conformance/offscreencanvas/context-lost-worker.html]: Same here

File: sdk/tests/conformance/offscreencanvas/context-lost-worker.html [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    finishTest();
+  }
+
+  var worker = new Worker('context-lost-worker.js');
+  worker.postMessage("Start worker");
+  worker.onmessage = function(e) {
+    if (e.data == "Test passed") {
+      testPassed("All tests have passed");
+      successfullyParsed = true;
```
Comment (zhenyao) [inline sdk/tests/conformance/offscreencanvas/context-lost-worker.html]: Same here

File: sdk/tests/conformance/offscreencanvas/context-lost.html [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+<script src="../../js/webgl-test-utils.js"></script>
+<script src="../../js/tests/canvas-tests-utils.js"></script>
+<script>
+function init()
+{
+    description("Tests behavior under a lost context for OffscreenCanvas");
+
+    var successfullyParsed = false;
+    if (!window.OffscreenCanvas) {
+      successfullyParsed = true;
```
Comment (zhenyao) [inline sdk/tests/conformance/offscreencanvas/context-lost.html]: Same here.

File: sdk/tests/conformance/offscreencanvas/context-lost.html [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+
+    // We need to initialize |uniformLocation| before losing context.
+    // Otherwise gl.getUniform() when context is lost will throw.
+    uniformLocation = gl.getUniformLocation(program, "tex");
+    extension.loseContext();
+
+    canvas.addEventListener("webglcontextlost", function() {
+        if (testLostContextWithoutRestore()) {
+            testPassed("All tests passed");
+            successfullyParsed = true;
```
Comment (zhenyao) [inline sdk/tests/conformance/offscreencanvas/context-lost.html]: Same here.

File: sdk/tests/conformance2/offscreencanvas/methods-2.html [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+</head>
+<body>
+<div id="description"></div>
+<div id="console"></div>
+<script>
+description("This test ensures that the WebGL context has all the methods in the specification.");
+
+var successfullyParsed = false;
+if (!window.OffscreenCanvas) {
+    successfullyParsed = true;
```
Comment (zhenyao) [inline sdk/tests/conformance2/offscreencanvas/methods-2.html]: Same here.

File: sdk/tests/conformance2/offscreencanvas/methods-2.html [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+    finishTest();
+}
+
+if (testAPIs('webgl2')) {
+  testPassed("All WebGL2 methods found");
+  successfullyParsed = true;
```
Comment (zhenyao) [inline sdk/tests/conformance2/offscreencanvas/methods-2.html]: Same here.

File: sdk/tests/conformance2/offscreencanvas/methods-2-worker.html [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+if (!window.OffscreenCanvas) {
+  finishTest();
+}
+
+var worker = new Worker('methods-2-worker.js');
+worker.postMessage("Start worker");
+worker.onmessage = function(e) {
+  if (e.data == "Test passed") {
+    testPassed("All test have passed");
+    successfullyParsed = true;
```
Comment (zhenyao) [inline sdk/tests/conformance2/offscreencanvas/methods-2-worker.html]: Same here

File: sdk/tests/conformance2/offscreencanvas/context-creation.html [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+</head>
+<body>
+<div id="description"></div>
+<div id="console"></div>
+<script>
+description("This test ensures that the WebGL2 context can be created on an OffscreenCanvas.");
+
+var successfullyParsed = false;
+if (!window.OffscreenCanvas) {
+  successfullyParsed = true;
```
Comment (zhenyao) [inline sdk/tests/conformance2/offscreencanvas/context-creation.html]: Same here.

File: sdk/tests/conformance2/offscreencanvas/context-creation.html [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+  finishTest();
+}
+
+if (contextCreation('webgl2')) {
+  testPassed("WebGL2 context created correctly.");
+  successfullyParsed = true;
```
Comment (zhenyao) [inline sdk/tests/conformance2/offscreencanvas/context-creation.html]: Same here.

File: sdk/tests/conformance2/offscreencanvas/context-creation-worker.html [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+  successfullyParsed = true;
+  finishTest();
+}
+
+var worker = new Worker('context-creation-worker.js');
+worker.postMessage("Start worker");
+worker.onmessage = function(e) {
+  if (e.data == "Test passed") {
+    testPassed("All tests have passed");
+    successfullyParsed = true;
```
Comment (zhenyao) [inline sdk/tests/conformance2/offscreencanvas/context-creation-worker.html]: Same here.

File: sdk/tests/conformance2/offscreencanvas/context-creation-worker.html [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+</head>
+<body>
+<div id="description"></div>
+<div id="console"></div>
+<script>
+description("This test ensures that the WebGL context can be created on an OffscreenCanvas.");
+
+var successfullyParsed = false;
+if (!window.OffscreenCanvas) {
+  successfullyParsed = true;
```
Comment (zhenyao) [inline sdk/tests/conformance2/offscreencanvas/context-creation-worker.html]: Same here

File: sdk/tests/conformance/offscreencanvas/methods.html [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+description("This test ensures that the WebGL context created for an OffscreenCanvas has all the methods in the specification.");
+
+var successfullyParsed = false;
+if (!window.OffscreenCanvas) {
+  successfullyParsed = true;
+  finishTest();
+}
+
+if (testAPIs('webgl')) {
+  successfullyParsed = true;
```
Comment (zhenyao) [inline sdk/tests/conformance/offscreencanvas/methods.html]: Same here.

File: sdk/tests/conformance/offscreencanvas/methods.html [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+</head>
+<body>
+<div id="description"></div>
+<div id="console"></div>
+<script>
+description("This test ensures that the WebGL context created for an OffscreenCanvas has all the methods in the specification.");
+
+var successfullyParsed = false;
+if (!window.OffscreenCanvas) {
+  successfullyParsed = true;
```
Comment (zhenyao) [inline sdk/tests/conformance/offscreencanvas/methods.html]: Same here

File: sdk/tests/conformance/offscreencanvas/methods-worker.html [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+  successfullyParsed = true;
+  finishTest();
+}
+
+var worker = new Worker("methods-worker.js");
+worker.postMessage("Start worker");
+worker.onmessage = function(e) {
+  if (e.data == "Test passed") {
+    testPassed("All tests have passed");
+    successfullyParsed = true;
```
Comment (zhenyao) [inline sdk/tests/conformance/offscreencanvas/methods-worker.html]: Same here.

File: sdk/tests/conformance/offscreencanvas/methods-worker.html [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+</head>
+<body>
+<div id="description"></div>
+<div id="console"></div>
+<script>
+description("This test ensures that the WebGL context created for an OffscreenCanvas has all the methods in the specification.");
+
+var successfullyParsed = false;
+if (!window.OffscreenCanvas) {
+  successfullyParsed = true;
```
Comment (zhenyao) [inline sdk/tests/conformance/offscreencanvas/methods-worker.html]: Same here

File: sdk/tests/conformance2/00_test_list.txt [hunk 25]
```diff
@@ -4,6 +4,7 @@ context/00_test_list.txt
 extensions/00_test_list.txt
 glsl3/00_test_list.txt
 misc/00_test_list.txt
+offscreencanvas/00_test_list.txt
```
Comment (zhenyao) [inline sdk/tests/conformance2/00_test_list.txt]: maybe say min-version 2.0.1?

Comment (kenrussell) [inline sdk/tests/conformance2/00_test_list.txt]: Yes, that's most appropriate now, just for bookkeeping purposes. (The 2.0.0 snapshot has already been taken under conformance-suites/ .)

**Comment by kenrussell** [comment]
Thanks @xidachen for refactoring. May I ask you to work with @zhenyao to finish this up and get it merged? I'm taking vacation the rest of this week. Thanks.

**Comment by xidachen** [comment]
@zhenyao Does that mean we don't even need the var successfullyParsed? We just need to do finishTest() at the end of the tests?

**Comment by zhenyao** [comment]
I think you still need to declare the successfullyParsed at the end.

**Comment by xidachen** [comment]
@zhenyao Changes have been made accordingly, please review. Thank you.

File: sdk/tests/conformance/offscreencanvas/context-creation-worker.html [hunk 26]
```diff
@@ -39,21 +39,18 @@
...
 var worker = new Worker('context-creation-worker.js');
 worker.postMessage("Start worker");
 worker.onmessage = function(e) {
   if (e.data == "Test passed") {
     testPassed("All tests have passed");
-    successfullyParsed = true;
   } else {
     testFailed("Some tests failed");
   }
+  var successfullyParsed = true;
```
Comment (zhenyao) [inline sdk/tests/conformance/offscreencanvas/context-creation-worker.html]: successfullyParsed should be declared at global scope, so right before </script> should be the right place.

Here and everywhere else.

**Comment by xidachen** [comment]
@zhenyao Thanks for pointing it out, I have made changes.

**Comment by zhenyao** [comment]
lgtm

**Comment by kenrussell** [comment]
Thanks @xidachen for doing this nice refactoring and @zhenyao for carrying through this code review. LGTM after the fact.

**Pull request merged**